pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -972,20 +972,33 @@ async def parse_packages(self, primary_xml, filelists_xml, other_xml, file_exten
         extra_repodata_parser = iterative_files_changelog_parser(
             file_extension, filelists_xml.path, other_xml.path
         )
+        written_pkgids = set()
         with ProgressReport(**progress_data) as packages_pb:
             while True:
                 try:
                     (pkgid, pkg) = packages.popitem(last=False)
                 except KeyError:
                     break
 
-                pkgid_extra, files, changelogs = next(extra_repodata_parser)
+                while True:
+                    pkgid_extra, files, changelogs = next(extra_repodata_parser)
+                    if pkgid_extra in written_pkgids:
+                        # This is a dirty hack to handle cases that "shouldn't" happen.
+                        # Sometimes repositories have packages listed twice under the same pkgid.
+                        # This is a problem because the primary.xml parsing deduplicates the
+                        # entries by placing them into a dict keyed by pkgid. So if we run into a
+                        # package that we've already written, we should skip it and move on.
+                        continue
+                    else:
+                        break
```

Comment:
Do I understand correctly that, originally primary and filelists/other were in the same order but with duplicates. E.g.
 ```
1   1
2   2
3   3
1   1
4   4
2   2
1   1
```
Now becasue primary is deduplicated by createrepo_c, it looks like this:
```
1   1
2   2 
3   3
4   1
    4
    2
    1
```

And you are still relying on the order, but you skip already seen items in the second column (filelists/other pkgid)?

I think I'm starting to be more concerned with the approach which is relying on the order. Maybe if we have another issue popped up in this area, we should reconsider the approach and do something close to what pulp 2 did.

Reply:
> And you are still relying on the order, but you skip already seen items in the second column (filelists/other pkgid)?

@goosemania Yes exactly

>we should reconsider the approach and do something close to what pulp 2 did.

I assume you mean parsing all 3 files ourselves?

Reply:
I mean the deduplication we do in pulp 2 for filelists/other.
I haven't checked the code recently but I thought we generate simple key-value dbs (on disk) for filelists and other and query info from there when walking through the primary.
I'm not saying it's the best solution but the idea is not to rely on the order but also not to load everything into memory. It probably does not require to parse primary ourselves.